### ISSUE TITLE
twisted.spread.jelly.unjelly breaks on classes defining __slots__

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4430]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| toidinamai reported|
|-|-|
|Trac ID|trac#4430|
|Type|defect|
|Created|2010-05-11 14:51:22Z|
|Branch|https://github.com/twisted/twisted/tree/jelly-slots-4430|

`twisted.spread.jelly.unjelly` breaks on classes that
do not have a `__dict__` attribute and define their
own `__setstate__` and `__getstate__` methods for serialization.
An example of those classes are those defining `__slots__`.

Attachments:

* [slots-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d75/d75883fb059312ccf4661f51578eb58fffa6e284/564f58ae497c604fbda85988d7080756255abf6f.patch) (1097 bytes) - added by toidinamai on 2010-05-11 14:54:46Z - new test case
* [slots-solution.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d75/d75883fb059312ccf4661f51578eb58fffa6e284/67adf0e1317ecf2ec5e3c62fec46325fe53a3728.patch) (1460 bytes) - added by toidinamai on 2010-05-11 14:55:11Z - implementation for the new test case

<details><summary>Searchable metadata</summary>

```
trac-id__4430 4430
type__defect defect
reporter__toidinamai toidinamai
priority__normal normal
milestone__ 
branch__branches_jelly_slots_4430 branches/jelly-slots-4430
branch_author__toidinamai toidinamai
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__pb__jelly____slots__ pb  jelly  __slots__
time__1273589482000000 1273589482000000
changetime__1273870253000000 1273870253000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="toidinamai's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|toidinamai removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [28928]) Branching to 'jelly-slots-4430'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [28929]) Apply toidinamai's patch

refs [#4430](https://github.com/twisted/twisted/issues/4430)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks for the patch.  A few simple things:

  1. `NO_STATE` should probably be private
  1. Operators should have whitespace on both sides.
  1. The docs for the `state` parameter to `_newInstance` should be updated.
  1. Methods should be separated from each other by two blank lines.
  1. Everything should have a docstring (particularly test methods)
  1. Copyright date on modified files should be updated
  1. There should be a news fragment describing the change

Nothing controversial here, so I'll make the changes and apply the patch.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [28937]) Merge jelly-slots-4430

Author: toidinamai
Reviewer: exarkun
Fixes: [#4430](https://github.com/twisted/twisted/issues/4430)

Support jellying and unjellying objects with no `__dict__` but which do
implement `__getstate__` and `__setstate__`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
